File: 2025-12-December-GFG-POTD/December 30 - Add Number Linked Lists/Solution.py
# class Node:
#     def __init__(self, data):
#         self.data = data
#         self.next = None

class Solution:
    # Function to add two numbers represented by linked lists.
    def addTwoLists(self, head1, head2):

        # Reverse both lists
        head1 = self.reverse(head1)
        head2 = self.reverse(head2)

        carry = 0
        dummy = Node(0)
        curr = dummy

        # Add digits
        while head1 or head2 or carry:
            total = carry

            if head1:
                total += head1.data
                head1 = head1.next

            if head2:
                total += head2.data
                head2 = head2.next

            carry = total // 10
            curr.next = Node(total % 10)
            curr = curr.next

        # Reverse result to restore correct order
        result = self.reverse(dummy.next)

        # Remove leading zeros
        while result and result.data == 0 and result.next:
            result = result.next

        return result

    # Helper function to reverse a linked list
    def reverse(self, head):
        prev = None
        curr = head

        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        return prev

File: 2025-12-December-GFG-POTD/December 30 - Add Number Linked Lists/Solution.cpp
/*
struct Node {
    int data;
    Node* next;

    Node(int d) {
        data = d;
        next = NULL;
    }
};
*/

class Solution {
public:
    Node* addTwoLists(Node* head1, Node* head2) {

        // Reverse both lists
        head1 = reverse(head1);
        head2 = reverse(head2);

        int carry = 0;
        Node* dummy = new Node(0);
        Node* curr = dummy;

        // Add digits
        while (head1 != NULL || head2 != NULL || carry != 0) {
            int sum = carry;

            if (head1 != NULL) {
                sum += head1->data;
                head1 = head1->next;
            }

            if (head2 != NULL) {
                sum += head2->data;
                head2 = head2->next;
            }

            carry = sum / 10;
            curr->next = new Node(sum % 10);
            curr = curr->next;
        }

        // Reverse result to restore correct order
        Node* result = reverse(dummy->next);

        // Remove leading zeros
        while (result != NULL && result->data == 0 && result->next != NULL) {
            result = result->next;
        }

        return result;
    }

private:
    // Helper function to reverse a linked list
    Node* reverse(Node* head) {
        Node* prev = NULL;
        Node* curr = head;

        while (curr != NULL) {
            Node* nextNode = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nextNode;
        }
        return prev;
    }
};
